### DISCUSSION TITLE
Merging the "unit" and "fuzz" tests under a new umbrella category "integration"

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#468] [Ideas]
A random idea just came to me.

What if we merged the "unit" and "fuzz" tests under a new umbrella category "integration"?

I know we used this name before, and we ended up separating them because we couldn't run them differently in CI. But that was before we started using the ["test_" and "testFuzz"](https://twitter.com/PaulRBerg/status/1616068157088714765) naming convention.

The current tests are not actually and rightfully "unit", as per my explanation [here](https://twitter.com/PaulRBerg/status/1347884079421812737).

**Comment by PaulRBerg** [comment]
I think we should proceed with my proposal and merge "fuzz" and "unit" into a single directory "integration" (or something else, e.g. "standard").

On top of my rationale above, I would now argue that it is a tad misleading to have "fuzz" as a testing category. It's more like an attribute, since the fork and the invariant tests fuzz the inputs, too.

Any thoughts, @andreivladbrg?

**Comment by andreivladbrg** [comment]
Sorry for my late response on this. Interesting idea, but I am not sure of this idea being the right choice. 

While I understand the appeal of this idea, especially given the improved code structure and reduction in lines of code, I remain uncertain about its suitability.

My concern lies in the theoretical correctness of the approach. Conceptually, I'm not convinced that consolidating these elements under the same umbrella is the appropriate choice.

- Integration tests are usually deterministic, meaning the same test will produce the same result. In contrast, fuzz testing is non-deterministic and uses automated methods to generate random inputs.
- Integration tests are used to validate that the system works correctly under expected conditions and to ensure components interact as intended. Fuzz tests, however, are designed to break the system (so are the invariants).

What we can do, and I agree with, is renaming the unit tests to integration.

**Reply by PaulRBerg** [reply]
No worries, thanks for your response.

You have used ChatGPT to write your response, and unfortunately this is a scenario in which ChatGPT can be more confusing than helpful. "Fuzzing" has a strange meaning in Foundry, different from what the industry typically means by this term. See https://github.com/foundry-rs/foundry/issues/4162.

> While I understand the appeal of this idea, especially given the improved code structure and reduction in lines of code, I remain uncertain about its suitability.

The wording here could have been better.

If you agree that the code structure is improved, you should consider the proposal as being "suitable".

> Integration tests are usually deterministic

Overly pedantic

> Fuzz tests, however, are designed to break the system

This is not quite right because you used ChatGPT and the Internet means something else by "fuzz testing". In Foundry, a fuzz test need not be designed to break the system, e.g.

```solidity
function testFuzz_Something(uint256 input) external {
    assertEq(foo.something(input), input + 1, "inputs not equal");
}
```

Nothing is "designed to be broken" in the test above, but this is a legitimate fuzz test in Foundry.

> What we can do, and I agree with, is renaming the unit tests to integration.

I have grown to have a strong opinion that we should unify the tests under a single umbrella, as I have done in #501.

Please reconsider the proposal.

You did not respond to my [argument](https://github.com/sablierhq/v2-core/discussions/468#discussioncomment-5936606) about fuzzing being an attribute rather than a category.

---

Separately, some feedback on your ChatGPT game: try to tweak the prompt and make it more informal. You sounded like a PhD student researching fuzz testing 😆

**Reply by andreivladbrg** [reply]
Yes, I have used GPT, but only to improve the points I was trying to make. I agree that it sounds too formal (you know me personally, and you know that I don't adhere to this method when talking with colleagues and close ones). I was in a hurry when I wrote it, and I didn't have the time to rephrase it to sound more casual.

> If you agree that the code structure is improved, you should consider the proposal as being "suitable".

I don't consider the proposal as "suitable" because this improvement also introduces an incorrect design pattern, which, IMO is more important.

> Overly pedantic 

I really don't agree. I consider it a major con to merge unit and fuzz under the same umbrella.

> test need not be designed to break the system, e.g
> Nothing is "designed to be broken" in the test above, but this is a legitimate fuzz test in Foundry

"Breaking the system" is one way of putting it. The point is that you want to generate inputs (ones that you couldn't think of during your unit tests) that may cause different behavior than the system would normally exhibit.
Regarding your example, I would say it is overly simplistic. Fuzz tests are most effective when applied to complex systems/flows, which we have in our codebase.

> Please reconsider the proposal.

I did reconsider it, and my opinion remains that placing them under the same umbrella is not the correct decision.

> you did not respond to my https://github.com/sablierhq/v2-core/discussions/468#discussioncomment-5936606 about fuzzing being an attribute rather than a category.

I don't think I agree. I would say that it is indeed a category. What I would consider as an attribute is the randomness with which the inputs are generated.

**Reply by PaulRBerg** [reply]
> I was in a hurry when I wrote it, and I didn't have the time to rephrase it to sound more casual.

This is absolutely fine; the point about ChatGPT was just FYI.

> introduces an incorrect design pattern

It is not incorrect, according to my arguments above.

> I consider it a major con to merge unit and fuzz under the same umbrella

Explanation-less comment.

> "Breaking the system" is one way of putting it.

Moving the goalposts

> Regarding your example, I would say it is overly simplistic

It was a simple example, but it was enough to refute your point that fuzz tests are "designed to break the system"

> Fuzz tests are most effective when applied to complex systems/flows, which we have in our codebase.

Moot point; we're not talking about "effectiveness" in this discussion.

> my opinion remains that placing them under the same umbrella is not the correct decision

Another explanation-less comment; you have yet to respond to my arguments above.

> I would say that it is indeed a category

It's a lousy category, because the invariants and the fork tests **subsume** it. Let me make a simple analogy ("sweet" is "fuzz"):

- Sweet
- Apples
  -  Sweet
- Snickers
  - Sweet
  
Both apples and Snickers bars are sweet, which makes "sweet" an _attribute_  - not a _category_!

> What I would consider as an attribute is the randomness with which the inputs are generated.

This is what fuzzing means in Foundry. Again, you are confusing fuzzing in general with fuzzing in the context of Foundry.

**Reply by PaulRBerg** [reply]
The big, big argument in favor of unifying "unit" and "fuzz" is avoiding the confusion discussed in https://github.com/foundry-rs/foundry/issues/4162.

What Foundry calls "invariants" is called "fuzzing" in other programming paradigms (even within Ethereum, e.g. in [Echidna](https://github.com/crytic/echidna)!). This makes placing a directory called "fuzz" next to one called "invariant" extremely confusing.

I have a very strong position on this issue, and I would like you to reconsider my reasons(s) for this.

**Comment by PaulRBerg** [comment]
@andreivladbrg - we need to make a decision about this.

Could you please read my arguments above, and either accept them (and change your mind) or refute them?

**Reply by andreivladbrg** [reply]
Given our time constraints it's not feasible for me to prolong this discussion.

Although I am not convinced that this is the right choice, I "accept" the change.

**Reply by PaulRBerg** [reply]
It is feasible. Please do continue the discussion, discuss rationally, and refute my points - if you can.

Otherwise, accept the proposal, and let's move on. But accept it because you _understand_, please.
